File: training/train.js
// const tf = require('@tensorflow/tfjs-node');
// const fs = require('fs');
// const csv = require('csv-parser');

// const irisData = [];
// const gazeData = [];

// fs.createReadStream('gaze_data.csv')
//   .pipe(csv())
//   .on('data', (row) => {
//     irisData.push([
//       parseFloat(row.left_iris_x),
//       parseFloat(row.left_iris_y),
//       parseFloat(row.right_iris_x),
//       parseFloat(row.right_iris_y)
//     ]);
//     gazeData.push([parseFloat(row.gaze_x), parseFloat(row.gaze_y)]);
//   })
//   .on('end', async () => {
//     console.log(`Loaded ${irisData.length} samples`);

//     const xs = tf.tensor2d(irisData);
//     const ys = tf.tensor2d(gazeData);

//     const model = tf.sequential();
//     model.add(tf.layers.dense({ inputShape: [4], units: 32, activation: 'relu' }));
//     model.add(tf.layers.dense({ units: 32, activation: 'relu' }));
//     model.add(tf.layers.dense({ units: 2 }));

//     model.compile({ optimizer: 'adam', loss: 'meanSquaredError' });

//     console.log("Training...");
//     await model.fit(xs, ys, {
//       epochs: 50,
//       batchSize: 32,
//       callbacks: {
//         onEpochEnd: (epoch, logs) => {
//           console.log(`Epoch ${epoch + 1}: loss = ${logs.loss.toFixed(4)}`);
//         }
//       }
//     });

//     // Ensure the directory exists
//     const modelDir = './model';
//     if (!fs.existsSync(modelDir)) {
//       fs.mkdirSync(modelDir, { recursive: true });
//     }

//     // Save the model with 'file://' prefix
//     await model.save(`file://${modelDir}/model.json`);
//     console.log("Model saved successfully!");
//   });

// const tf = require('@tensorflow/tfjs-node');
// const fs = require('fs');
// const csv = require('csv-parser');

// const irisData = [];
// const gazeData = [];

// fs.createReadStream('gaze_data.csv')
//   .pipe(csv())
//   .on('data', (row) => {
//     // Skip rows with missing data (essential for training)
//     if (
//       row.left_iris_x && row.left_iris_y &&
//       row.right_iris_x && row.right_iris_y &&
//       row.gaze_x && row.gaze_y
//     ) {
//       irisData.push([
//         parseFloat(row.left_iris_x),
//         parseFloat(row.left_iris_y),
//         parseFloat(row.right_iris_x),
//         parseFloat(row.right_iris_y)
//       ]);

//       // Normalize gaze_x and gaze_y from 0–1 (if your grid is 20x20 or resolution-based)
//       gazeData.push([
//         parseFloat(row.gaze_x) / 20,   // 🔁 Normalize if using 20x20 grid
//         parseFloat(row.gaze_y) / 20
//       ]);
//     }
//   })
//   .on('end', async () => {
//     console.log(`Loaded ${irisData.length} samples`);

//     const xs = tf.tensor2d(irisData);
//     const ys = tf.tensor2d(gazeData);

//     // Model definition
//     const model = tf.sequential();
//     model.add(tf.layers.dense({ inputShape: [4], units: 32, activation: 'relu' }));
//     model.add(tf.layers.dense({ units: 32, activation: 'relu' }));
//     model.add(tf.layers.dense({ units: 2 })); // Output is [gaze_x, gaze_y]

//     model.compile({
//       optimizer: tf.train.adam(),
//       loss: 'meanSquaredError'
//     });

//     console.log("Training...");
//     await model.fit(xs, ys, {
//       epochs: 100,
//       batchSize: 32,
//       callbacks: {
//         onEpochEnd: (epoch, logs) => {
//           console.log(`Epoch ${epoch + 1}: loss = ${logs.loss.toFixed(5)}`);
//         }
//       }
//     });

//     // Create model directory if not exists
//     const modelDir = './model';
//     if (!fs.existsSync(modelDir)) {
//       fs.mkdirSync(modelDir, { recursive: true });
//     }

//     // Save trained model to filesystem
//     await model.save(`file://${modelDir}/model.json`);
//     console.log("✅ Model saved successfully in ./model folder");
//   });

const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');
const csv = require('csv-parser');

// Will read screen width/height from data instead of hardcoding
const irisData = [];
const gazeData = [];
const screenSizes = [];

class EarlyStopping {
    constructor({ monitor = 'val_loss', patience = 0, minDelta = 0 }) {
        this.monitor = monitor;
        this.patience = patience;
        this.minDelta = minDelta;
        this.best = Infinity;
        this.patienceCounter = 0;
        this.stopTraining = false;
    }

    onEpochEnd(epoch, logs) {
        const current = logs[this.monitor];
        if (current < this.best - this.minDelta) {
            this.best = current;
            this.patienceCounter = 0;
        } else {
            this.patienceCounter += 1;
            if (this.patienceCounter >= this.patience) {
                console.log(
                    `Early stopping: No improvement in ${this.monitor} for ${this.patience} epochs`
                );
                this.stopTraining = true;
            }
        }
    }
}

// Shuffle both arrays in sync
function shuffleInUnison(a, b, c) {
    for (let i = a.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [a[i], a[j]] = [a[j], a[i]];
        [b[i], b[j]] = [b[j], b[i]];
        [c[i], c[j]] = [c[j], c[i]];
    }
}

fs.createReadStream('gaze_data.csv')
    .pipe(csv())
    .on('data', (row) => {
        const sw = parseFloat(row.screen_width);
        const sh = parseFloat(row.screen_height);
        const gx = parseFloat(row.gaze_x);
        const gy = parseFloat(row.gaze_y);

        if (
            row.left_iris_x && row.left_iris_y &&
            row.right_iris_x && row.right_iris_y &&
            gx >= 0 && gy >= 0 &&
            sw > 0 && sh > 0
        ) {
            irisData.push([
                parseFloat(row.left_iris_x),
                parseFloat(row.left_iris_y),
                parseFloat(row.right_iris_x),
                parseFloat(row.right_iris_y)
            ]);
            gazeData.push([gx / sw, gy / sh]); // normalized target
            screenSizes.push([sw, sh]);
        }
    })
    .on('end', async () => {
        console.log(`Loaded ${irisData.length} samples`);
        if (irisData.length === 0) return;

        // Shuffle before creating tensors
        shuffleInUnison(irisData, gazeData, screenSizes);

        const xs = tf.tensor2d(irisData);
        const ys = tf.tensor2d(gazeData);

        const model = tf.sequential();
        model.add(tf.layers.dense({ inputShape: [4], units: 64, activation: 'relu' }));
        model.add(tf.layers.dropout({ rate: 0.2 }));
        model.add(tf.layers.dense({ units: 64, activation: 'relu' }));
        model.add(tf.layers.dropout({ rate: 0.2 }));
        model.add(tf.layers.dense({ units: 32, activation: 'relu' }));
        model.add(tf.layers.dense({ units: 2 }));

        model.compile({
            optimizer: tf.train.adam(0.001),
            loss: 'meanSquaredError'
        });

        const earlyStopping = new EarlyStopping({
            monitor: 'val_loss',
            patience: 10,
            minDelta: 0.0001
        });

        console.log("Training...");
        await model.fit(xs, ys, {
            epochs: 200,
            batchSize: 32,
            validationSplit: 0.2,
            callbacks: {
                onEpochEnd: (epoch, logs) => {
                    console.log(
                        `Epoch ${epoch + 1}: loss=${logs.loss.toFixed(5)}, val_loss=${logs.val_loss.toFixed(5)}`
                    );
                    earlyStopping.onEpochEnd(epoch, logs);
                },
                onBatchEnd: () => {
                    if (earlyStopping.stopTraining) return { stop: true };
                }
            }
        });

        // Test a sample: convert normalized prediction back to pixels
        const testSample = tf.tensor2d([irisData[0]]);
        const prediction = model.predict(testSample);
        const [predXNorm, predYNorm] = prediction.dataSync();

        const [sw, sh] = screenSizes[0];
        const predX = predXNorm * sw;
        const predY = predYNorm * sh;

        console.log("Predicted gaze (pixels):", [predX, predY]);
        console.log("Actual gaze (pixels):", [gazeData[0][0] * sw, gazeData[0][1] * sh]);

        const modelDir = './model';
        if (!fs.existsSync(modelDir)) fs.mkdirSync(modelDir, { recursive: true });
        await model.save(`file://${modelDir}/model.json`);
        console.log("✅ Model saved successfully in ./model folder");

        xs.dispose();
        ys.dispose();
        testSample.dispose();
        prediction.dispose();
    });

File: training/train.py
import pandas as pd
import numpy as np
import os
from sklearn.utils import shuffle
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Load and process data (same as above, with feature engineering)
iris_data = []
gaze_data = []
screen_sizes = []

df = pd.read_csv('gaze_data.csv')
for _, row in df.iterrows():
    sw = float(row['screen_width'])
    sh = float(row['screen_height'])
    tx = float(row['target_x'])
    ty = float(row['target_y'])

    if (pd.notna(row['left_iris_x']) and pd.notna(row['left_iris_y']) and
        pd.notna(row['right_iris_x']) and pd.notna(row['right_iris_y']) and
        tx >= 0 and ty >= 0 and sw > 0 and sh > 0):
        left_x, left_y = float(row['left_iris_x']), float(row['left_iris_y'])
        right_x, right_y = float(row['right_iris_x']), float(row['right_iris_y'])

        dx = right_x - left_x
        dy = right_y - left_y
        mid_x = (left_x + right_x) / 2
        mid_y = (left_y + right_y) / 2

        iris_data.append([left_x, left_y, right_x, right_y, dx, dy, mid_x, mid_y])
        gaze_data.append([tx / sw, ty / sh])
        screen_sizes.append([sw, sh])

print(f"Loaded {len(iris_data)} samples")
if not iris_data:
    exit()

# Shuffle and normalize
iris_data, gaze_data, screen_sizes = shuffle(iris_data, gaze_data, screen_sizes, random_state=42)
scaler = MinMaxScaler()
iris_data = scaler.fit_transform(iris_data)

# Split
X_train, X_test, y_train, y_test, screens_train, screens_test = train_test_split(
    iris_data, gaze_data, screen_sizes, test_size=0.3, random_state=42
)

# Train RandomForest
model = RandomForestRegressor(n_estimators=200, max_depth=15, random_state=42)
model.fit(X_train, y_train)

# Predict and evaluate on full test set
predictions = model.predict(X_test)
errors = []
within_50 = 0
within_100 = 0

for i in range(len(X_test)):
    pred_x_norm, pred_y_norm = predictions[i]
    sw, sh = screens_test[i]
    pred_x = pred_x_norm * sw
    pred_y = pred_y_norm * sh
    actual_x = y_test[i][0] * sw
    actual_y = y_test[i][1] * sh

    error = np.sqrt((pred_x - actual_x) ** 2 + (pred_y - actual_y) ** 2)
    errors.append(error)

    if error <= 50:
        within_50 += 1
    if error <= 100:
        within_100 += 1

    if i < 20:
        print(f"Test Sample {i + 1}: Pred=({pred_x:.2f}, {pred_y:.2f}), Actual=({actual_x:.2f}, {actual_y:.2f}), Error={error:.2f} px")

# from sklearn.model_selection import cross_val_score
# scores = cross_val_score(model, iris_data, gaze_data, cv=5, scoring='neg_mean_squared_error')
# print("Cross-validated MSE:", -np.mean(scores))

mean_error = np.mean(errors)
print(f"\n📊 Mean error over {len(errors)} test samples: {mean_error:.2f} px")
print(f"✅ % within 50 px: {(within_50 / len(errors) * 100):.1f}%")
print(f"✅ % within 100 px: {(within_100 / len(errors) * 100):.1f}%")

# Save model (using joblib)
import joblib
model_dir = './model_rf'
os.makedirs(model_dir, exist_ok=True)
joblib.dump(model, os.path.join(model_dir, 'model_rf.pkl'))
print("✅ Model saved successfully in ./model_rf folder")